cider/function/substrait/SubstraitExtension.cpp

```diff
@@ -0,0 +1,349 @@
+      "functions_aggregate_approx.yaml",
+      "functions_aggregate_generic.yaml",
+      "functions_arithmetic.yaml",
+      "functions_arithmetic_decimal.yaml",
+      "functions_boolean.yaml",
+      "functions_comparison.yaml",
+      "functions_datetime.yaml",
+      "functions_logarithmic.yaml",
+      "functions_rounding.yaml",
+      "functions_string.yaml",
+      "functions_set.yaml",
+      "unknown.yaml",
+  };
+  const auto& extensionRootPath = getSubstraitExtensionAbsolutePath();
+  return loadExtension(extensionRootPath, extensionFiles);
+}
+
+std::shared_ptr<SubstraitExtension> SubstraitExtension::loadExtension(
+    const std::string& basePath,
+    const std::vector<std::string>& extensionFiles) {
+  std::vector<std::string> yamlExtensionFiles;
+  yamlExtensionFiles.reserve(extensionFiles.size());
+  for (auto& extensionFile : extensionFiles) {
+    auto const pos = basePath.find_last_of('/');
+    const auto& extensionUri = basePath.substr(0, pos) + "/" + extensionFile;
+    yamlExtensionFiles.emplace_back(extensionUri);
+  }
+  return loadExtension(yamlExtensionFiles);
+}
+
+std::shared_ptr<SubstraitExtension> SubstraitExtension::loadExtension(
+    const std::vector<std::string>& yamlExtensionFiles) {
+  SubstraitExtension mergedExtension;
+  for (const auto& extensionUri : yamlExtensionFiles) {
+    const auto& substraitExtension =
+        YAML::LoadFile(extensionUri).as<SubstraitExtension>();
+
+    for (auto& scalarFunctionVariant : substraitExtension.scalarFunctionVariants) {
+      scalarFunctionVariant->uri = extensionUri;
+      mergedExtension.scalarFunctionVariants.emplace_back(scalarFunctionVariant);
+    }
+
+    for (auto& aggregateFunctionVariant : substraitExtension.aggregateFunctionVariants) {
+      aggregateFunctionVariant->uri = extensionUri;
+      mergedExtension.aggregateFunctionVariants.emplace_back(aggregateFunctionVariant);
+    }
+
+    for (auto& type : substraitExtension.types) {
+      type->uri = extensionUri;
+      mergedExtension.types.emplace_back(type);
+    }
+  }
+  return std::make_shared<SubstraitExtension>(mergedExtension);
+}
+
+std::optional<SubstraitFunctionVariantPtr> SubstraitExtension::lookupScalarFunction(
+    const std::string& signature) const {
+  for (const auto& scalarFunctionVariant : scalarFunctionVariants) {
+    if (scalarFunctionVariant->signature() == signature) {
+      return std::make_optional(scalarFunctionVariant);
+    }
+  }
+  return std::nullopt;
+}
+
+std::optional<SubstraitFunctionVariantPtr> SubstraitExtension::lookupAggregateFunction(
+    const std::string& signature) const {
+  for (const auto& aggregateFunctionVariant : aggregateFunctionVariants) {
+    if (aggregateFunctionVariant->signature() == signature) {
+      return std::make_optional(aggregateFunctionVariant);
+    }
+  }
+  return std::nullopt;
+}
+
+std::optional<SubstraitFunctionVariantPtr> SubstraitExtension::lookupFunction(
+    const std::string& signature) const {
```

Comment:
We should call these API which contains the match rule with input types.

Reply:
This is velox's function lookup, the signatue such as: add:int_int contains the input types.